### PR TITLE
Step 6: Add clang cl pipeline for Windows

### LINKED ISSUE #11974: Windows builds with clang-cl.exe
In order to reduce the need to engineer workarounds when interoperating with projects such as Chromium that build with clang-cl, and to provide greater consistency with the rest of our Clang-based builds, we should use clang-cl.exe instead of cl.exe. Both are compatible with MSVC.

The main blocker here appears to be https://github.com/google/googletest/issues/2490.

### PR DESCRIPTION
**Pull request by wrowe**
Commit Message: Step 6: Add clang cl pipeline for Windows
Additional Description:

Now that we have regenerated envoy build tools to update bazel and llvm, and picked up the llvm 11 tooling for windows on envoyproxy/master, finally introduce a clang-cl pipeline based on clang-cl

Co-authored-by: Sunjay Bhatia sunjayb@vmware.com

Risk Level: low (new facility)
Testing: local
Docs Changes: TBD
Release Notes: TBD
Platform Specific Features: Windows clang-cl compilation
Replaces https://github.com/envoyproxy/envoy/pull/14135
Fixes #11974

### CODE REVIEW

**Review by sunjayBhatia — CHANGES_REQUESTED** [review]
Could you add the repository cache task that was recently merged to the clang-cl pipeline: https://github.com/envoyproxy/envoy/pull/14678/files#diff-8ca122481a2ae193003b523bf5b7a9cfa0e914be502d483b6adc1d06668c724a

This should help with download flakiness on Windows

File: .azure-pipelines/pipelines.yml
Comment (wrowe) [inline .azure-pipelines/pipelines.yml:426]: @sunjayBhatia do we want to use the same key here as the other build, e.g. share a cache, or does the clang-cl cache need to be different? I'm guessing if they are the repository objects, sharing them is desired?

Comment (sunjayBhatia) [reply .azure-pipelines/pipelines.yml:426]: Thats a good question, I'm not sure if anything build specific goes in there but as you say, makes sense if it didn't

@lizan Do you have any reasoning behind why the different Linux jobs have separate cache entries? https://github.com/envoyproxy/envoy/blob/7653adfd88379586cc55518b0548ba88fbea6dae/.azure-pipelines/bazel.yml#L24

Comment (wrowe) [reply .azure-pipelines/pipelines.yml:426]: I'll leave this as-is, a distinct cache for clang-cl, pending lizan's feedback.

Comment (lizan) [reply .azure-pipelines/pipelines.yml:426]: I split this because different build targets are pulling slightly different set of repositories, because repositories are lazily pulled. For example the compile_time_options will pull boringssl_fips instead of normal boringssl, additional wasm runtimes such as wasmtime/wavm.

So if clang-cl and msvc-cl shares exact same set of repositories, they don't have to be a different key, if not we can keep them different. Up to you two.

File: .azure-pipelines/pipelines.yml
Comment (wrowe) [inline .azure-pipelines/pipelines.yml:419]: Question to @lizan and @envoyproxy/windows-dev maintainers; do we want to wait for msvc-cl "release" before beginning the clang-cl build? Or are there typically few enough modified components that we would have enough workers for both compilations at once, with minimal delay in seeing a release and then a docker result?

Comment (lizan) [reply .azure-pipelines/pipelines.yml:419]: I prefer having them in parallel.

Comment (wrowe) [reply .azure-pipelines/pipelines.yml:419]: Done

**Comment by wrowe** [comment]
@lizan , @sunjayBhatia this looks ready for final review.

File: .azure-pipelines/pipelines.yml
Comment (sunjayBhatia) [inline .azure-pipelines/pipelines.yml:420]: Any specific reason this was removed? b/c of image sizes I think this job can take longer than the default 60 minutes

File: ci/windows_ci_steps.sh
Comment (sunjayBhatia) [inline ci/windows_ci_steps.sh:94]: Does the quoting not work here to collapse these two cases into one line?

**Review by sunjayBhatia — COMMENTED** [review]
Just a couple questions, otherwise LGTM

File: ci/windows_ci_steps.sh
Comment (wrowe) [reply ci/windows_ci_steps.sh:94]: Because of the outer single-tick quotes, starting at `'kind`, the inner expansion would never occur, which is the actually useful one.

File: .azure-pipelines/pipelines.yml
Comment (wrowe) [reply .azure-pipelines/pipelines.yml:420]: No, not deliberately, I can correct.

File: ci/windows_ci_steps.sh
Comment (wrowe) [reply ci/windows_ci_steps.sh]: I'm dropping the logic, we aren't paying attention to this, and are directly examining the few remaining tests. I left the example you drafted, in this case, the skip pattern. Marking resolved.

**Comment by wrowe** [comment]
Once again, @sunjayBhatia and @lizan, I think all the concerns have been asked and answered, but let me know if I can add any more polish to this submission before it's merged. We will be iterating over .pdb output (just as on msvc-cl) and the build output (right now, a 33% larger binary than the equivalent MS tooling, speed delta TBD.)

**Review by sunjayBhatia — APPROVED** [review]

**Comment by wrowe** [comment]
@htuch as this is your ask, and it's been reviewed and lizan's comments addressed, if you would like to move ahead please go ahead and merge this.

**Comment by htuch** [comment]
Amazing, thank you!

**Review by htuch — APPROVED** [review]

**Pull request merged**
